### PR TITLE
Refactoring Codebase

### PR DESCRIPTION
**Pull request by mzfr**
I have refactored the code and have broken the functions into different files. I have made so many files because I think this will increase the maintainability.
For example handle_files.py file contains all the functions that will never have to be touched again.

This is the new structure of our code base:

### Structure

* handle_files.py (This deals with file, find any file or create multiple files)
    - find_file()
    - find_file_recursive()
    - create_file_index()
    - find_in_file
    - addon_file_exists (search for addon file only)

* Check_files.py (checks all the files/folders like addon.xml,any json file )
    - check_for_invalid_xml
    - check_for_invalid_json
    - check_addon_xml
    - addon_xml_matches_folder
    - check_for_legacy_language_path
    - check_file_whitelist

* check_artwork.py (checks all the images, artworks or screenshot)
    - check_artwork
    - _check_image_type
    - _assests
    - _check_icon
    - _check_fanart

* check_string.py (check for particular terms which are blacklisted)
    - check_for_legacy_strings_xml
    - find_blacklisted_strings

* check_entrypoint.py (check for complex entry point)
    - check_complex_addon_entrypoint
    - _number_of_lines

* check_dependency.py (check other dependency of addon)
    - check_addon_dependencies
    - _get_users_dependencies

* check_addon.py
    - start
    - all_repo_addons
    - _get_addons

* common.py
    - has_transperency
    - load_plugins
    - relative_path

* check_old_addon.py
    - check_for_existing_addon
    - _get_addon_name

After refactoring I tested the complexity and other things with pylint and radon. Here's some stats after and before refactoring.

### Stats

* raw metrics
```
            After       Before
    LOC:    1014        927
    LLOC:   588         570
    SLOC:   654         649
    Multi:  152         70
    Blank:  186         178
```

* Complexity
```
                        After                     Before
    average =           3.090909090909091         3.292307692307692
    _check_image_type   C(11)                     D (25)
```

* maintainibility:
```
                            After               Before

    logger.py -             A (100.00)          A (100.00)
    common.py -             A (55.61)           A (57.55)
    check_repo.py -         A (66.26)           A (66.26)
    check_addon.py -        A (61.07)           A (23.54)
    __main__.py -           A (55.60)           A (55.60)
    check_old_addon.py -    A (59.29)
    check_string.py -       A (64.71)
    check_files.py -        A (54.00)
    check_entrypoint.py -   A (53.21)
    handle_files.py -       A (46.91)
    check_dependency.py -   A (54.72)
    check_artwork.py -      A (51.12)
```

* pylint score
```
        After           Before
        9.95/10         9.60/10
```

**NOTE**

* while testing with pylint I disabled all the C(convention) and R(refactor) errors because with those modes on pylint haven to gives suggestion of itself. But this was done for both the case i.e after and before refactoring.

* In complexity i have only mentioned about a single function because other function were under 10 even before refactoring.

* I tried my best to name files accordingly so if you think the files must be rename please let me knoe.

### CODE REVIEW

**Comment by Razzeee** [comment]
Approach seems fine. We should run this against a complete repo and compare the reported output old vs new.

Other then that it just needs a review.

**Comment by mzfr** [comment]
Okay so I'll run it again repo-plugin(Jarvis/krypton) and will post the output

**Comment by mzfr** [comment]
Test is failing because now we are passing 5 argument instead of 4.

**Comment by Rechi** [comment]
Then why not adjust the tests too?

This time I would make multiple commits , for example
- refactoring: move artwork check into separate file
- refactoring: move dependencies check into separate file
- refactoring: move entrypoint check into separate file
- refactoring: move files check into separate file
- refactoring: move old_addon check into separate file
- refactoring: move string check into separate file
- catch exceptions per addon for a repo check
- add PR argument

**Comment by mzfr** [comment]
@Rechi okay so Should I close this PR ? and make a new Pull

**Comment by Rechi** [comment]
`.travis-repo.yml.example` also has to be adjusted to pass the new `--PR` argument for PR builds.

Just update this PR.

File: kodi_addon_checker/__main__.py [hunk 1]
```diff
@@ -75,9 +76,12 @@ def main():
         # Following report is a wrapper for all sub reports
         report = Report("")
         for directory in args.dir:
-            report.add(check_artifact(directory, args, args.branch))
+            if args.PR:
```
Comment (Rechi) [inline kodi_addon_checker/__main__.py]: why is this if needed?

File: kodi_addon_checker/__main__.py [hunk 2]
```diff
@@ -75,9 +76,12 @@ def main():
...
+                report.add(check_artifact(directory, args, args.branch, args.PR))
+            else:
+                report.add(check_artifact(directory, args, args.branch, False))
     else:
-        report = check_artifact(os.getcwd(), args, args.branch)
+        report = check_artifact(os.getcwd(), args, args.branch, False)
```
Comment (Rechi) [inline kodi_addon_checker/__main__.py]: why hardcode `False` here?

File: kodi_addon_checker/check_repo.py [hunk 3]
```diff
@@ -4,15 +4,19 @@
...
     for addon_folder in toplevel_folders:
         if addon_folder[0] != '.':
             addon_path = os.path.join(repo_path, addon_folder)
-            addon_report = check_addon.start(addon_path, branch_name, all_repo_addons, config)
-            repo_report.add(addon_report)
+            try:
+                addon_report = check_addon.start(addon_path, branch_name, all_repo_addons, PR, config)
+                repo_report.add(addon_report)
+
+            except IOError:
```
Comment (Rechi) [inline kodi_addon_checker/check_repo.py]: why only catch `IOError`?

File: kodi_addon_checker/__main__.py [hunk 4]
```diff
@@ -29,29 +29,28 @@ def dir_type(dir_path):
     return os.path.abspath(dir_path)
 
 
-def check_artifact(artifact_path, args, branch_name):
+def check_artifact(artifact_path, args, branch_name, PR):
```
Comment (Razzeee) [inline kodi_addon_checker/__main__.py]: variables are snake case here. So lowercase `pr`

**Comment by Razzeee** [comment]
I would prefer this to be two PRs, as one part shouldn't affect behavior if we do it correctly. As it only refactors. And the other part does affect behavior.

File: kodi_addon_checker/check_old_addon.py [hunk 5]
```diff
@@ -21,7 +21,13 @@ def check_for_existing_addon(report: Report, addon_path: str, all_repo_addons: d
...
 
         if addon_name in repo_addons:
-            _check_versions(report, addon_name, branch, addon_version, repo_addons[addon_name], PR)
+            if LooseVersion(addon_version) <= LooseVersion(repo_addons[addon_name]):
+                report.add(Record(PROBLEM, "%s addon already exist with version: %s in %s branch"
+                                  % (addon_name, repo_addons[addon_name], branch)))
+            return
+        else:
+            report.add(Record(INFORMATION, "%s addon also exists in %s branch but with version: %s"
+                              % (addon_name, branch, repo_addons[addon_name])))
```
Comment (Rechi) [inline kodi_addon_checker/check_old_addon.py]: something went wrong here during the rebase

**Comment by Rechi** [comment]
@mzfr please create a separate PR for catching exceptions per plugin for a repo check

**Comment by Razzeee** [comment]
Can you please do a diff for the report of this vs the code code before again?

**Comment by mzfr** [comment]
# stats

* Raw metrics

```
                    Before      After
    LOC:            904         1052
    LLOC:           552         600
    SLOC:           635         677
    Comments:       31          26
    Multi:          62          150
    Blank:          172         195
```

* complexity

```                     Before                     After
    average =           3.2698412698412698         3.1363636363636362
    _check_image_type   D(25)                      C (11)

```

* maintainibility:
```
                            After               Before

    logger.py -             A (100.00)          A (100.00)
    common.py -             A (55.61)           A (57.55)
    check_repo.py -         A (66.26)           A (66.26)
    check_addon.py -        A (65.99)           A (23.54)
    __main__.py -           A (72.53)           A (73.50)
    check_old_addon.py -    A (76.99)
    check_string.py -       A (93.74)
    check_files.py -        A (70.11)
    check_entrypoint.py -   A (78.96)
    handle_files.py -       A (76.04)
    check_dependency.py -   A (78.74)
    check_artwork.py -      A (55.32)
```

* Pylint:
```
    Before      9.69/10
    After       9.91/10
```

**Comment by Rechi** [comment]
Please rebase this correctly, `Remove exception for a repo check` commit shouldn't be in this PR.

**Comment by mzfr** [comment]
I am not sure what to use for commit message because in this last commit message I have actually removed the old exception code so that is why I used `remove exception code`.

Can you please show me the error message ?

File: kodi_addon_checker/check_old_addon.py [hunk 6]
```diff
@@ -51,3 +42,12 @@ def _check_versions(report: Report, addon_name, branch, addon_version, repo_addo
...
                               % (addon_name, branch, repo_addons_version)))
+
+
+def _get_addon_name(xml_path: str):
+    """returns name and version of the addon
+        :xml_path: path of the xml file
+    """
+
+    tree = ET.parse(xml_path).getroot()
+    return (tree.get("id"), tree.get("version"))
```
Comment (Rechi) [inline kodi_addon_checker/check_old_addon.py]: reordering has to be squashed into the commit which adds check_old_addon.py

File: kodi_addon_checker/check_addon.py [hunk 7]
```diff
@@ -1,20 +1,19 @@
...
 from kodi_addon_checker.report import Report
+from kodi_addon_checker.record import Record, INFORMATION
 from kodi_addon_checker import check_artwork
 from kodi_addon_checker import check_old_addon
 from kodi_addon_checker import check_dependencies
 from kodi_addon_checker import check_entrypoint
-from kodi_addon_checker import handle_files
 from kodi_addon_checker import check_files
 from kodi_addon_checker import check_string
+from kodi_addon_checker import handle_files
```
Comment (Rechi) [inline kodi_addon_checker/check_addon.py]: reordering has to be squashed into the commit which adds handle_files

File: kodi_addon_checker/check_addon.py [hunk 8]
```diff
@@ -1,20 +1,19 @@
 import os
-import xml.etree.ElementTree as ET
```
Comment (Rechi) [inline kodi_addon_checker/check_addon.py]: unrelated

File: kodi_addon_checker/check_addon.py [hunk 9]
```diff
@@ -1,20 +1,19 @@
...
 import requests
 import logging
-from kodi_addon_checker import logger
```
Comment (Rechi) [inline kodi_addon_checker/check_addon.py]: unrelated

File: kodi_addon_checker/check_addon.py [hunk 10]
```diff
@@ -1,20 +1,19 @@
...
 import gzip
 from io import BytesIO
 
-from kodi_addon_checker.record import Record, INFORMATION
```
Comment (Rechi) [inline kodi_addon_checker/check_addon.py]: unrelated

File: kodi_addon_checker/check_addon.py (hunk 7 above)
Comment (mzfr) [reply kodi_addon_checker/check_addon.py]: how do i squash only one line change into an older commit? i've only used squash to merge two entire commits.

File: kodi_addon_checker/check_files.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        file_parts = file["name"].rsplit(".")
+        if len(file_parts) > 1:
+            file_ending = "." + file_parts[len(file_parts) - 1]
+            if re.match(whitelist, file_ending, re.IGNORECASE) is None:
+                report.add(Record(WARNING,
+                                  "Found non whitelisted file ending in filename %s" %
+                                  relative_path(os.path.join(file["path"], file["name"]))))
+                report.add(Record(
+                    WARNING, "Using the old language directory structure, please move to the new one."))
+                # found_warning = True
```
Comment (Rechi) [inline kodi_addon_checker/check_files.py]: `old language directory`  shouldn't be here
